content/blogposts/2022/ranking-in-a-week.md

```diff
@@ -0,0 +1,130 @@
+3,github.com/sourcegraph/blog-post-example,internal/database.go
+4,github.com/sourcegraph/blog-post-example,internal/experiments.go
+5,github.com/sourcegraph/blog-post-example,internal/handler.go
+6,github.com/sourcegraph/blog-post-example,internal/init.go
+```
+
+Next, we output `references.csv`, which corresponds to the edges _between_ the white vertices. Such an edge is output when there is a direct reference from a variable defined in another file.
+
+```csv
+from,to
+1,3
+1,4
+2,5
+3,4
+4,3
+4,5
+5,3
+5,4
+6,3
+6,5
+```
+
+To support references between repositories, we output `monikers.csv`, which will declare a set of _monikers_ that are either declared in or referenced by a particular path. If one repository declares an `import` moniker, and another repository declares an `export` moniker with the same scheme and identifier, then we create an edge between the documents across repositories.
+
+```csv
+pathID,monikerType,schema,identifier
+1,import,gomod,github.com/sourcegraph/log/Logger
+1,import,gomod,github.com/sourcegraph/log/Logger.Log
+5,import,gomod,github.com/sourcegraph/log/Logger
+5,import,gomod,github.com/sourcegraph/log/Logger.Log
+6,import,gomod,github.com/sourcegraph/log/Logger
+6,import,gomod,github.com/sourcegraph/log/Logger.Log
+```
+
+We would later round out this implementation by also [pruning irrelevant data](https://github.com/sourcegraph/sourcegraph/pull/43832) from the target GCS bucket as newer indexes for repositories replace older indexes. Since the landing of this export task, it has become the basis of several other solutions currently in-flight (details to come in future blog posts).
+
+By October 27th, we had a few successful runs of a distributed Spark job over the graph described by these CSV files and hand-vetted the document scores results to ensure our mental model and reality match. In order to build a better intuition of how PageRank will relate to source code documents, we constructed several variations of the graph:
+
+1. Edges are directional, and `a -> b` exists when `a` refers to a symbol defined in `b`
+1. Edges are directional, and `b -> a` exists when `a` refers to a symbol defined in `b`
+1. Edges are undirected, and both `a -> b` and `b -> a` exist when one document refers to a symbol defined in the other
+
+Interestingly, we ended up running PageRank over an undirected graph, as directional edges tended to rank auto-generated files with _tons_ of definitions (e.g., a protobuf code gen) very highly, but rank `main`-like files that refers to lots of definitions but defines few themselves very low. This is the opposite of what we want: the glue code with few inbound edges but lots of (unique) outbound edges indicates the collision of two program boundaries. These usages are precisely the ones that should be advertised, and using undirected edges gives us the best of both directional options by boosting files with multiple inbound _and_ outbound edges.
```

Comment:
Intuitively, I would have assumed that changing the direction of the edge, from `a -> b` to 'b -> a' would have inverted the situation IE led to higher ranking `main`-like files and lower ranking auto-generating files. However, as it is stated here, both types of directed graphs led to the same outcome?

Reply:
```suggestion
Interestingly, we ended up running PageRank over an undirected graph, as directional edges tended to rank auto-generated files with _tons_ of definitions (e.g., a protobuf code gen) very highly, but rank `main`-like files that refers to lots of definitions but defines few themselves very low. This is the opposite of what we want: the glue code with few inbound edges but lots of (unique) outbound edges indicates the collision of two program boundaries. These usages are precisely the ones that should be advertised, and using undirected edges gives us the best of both directional options by boosting files with edges (inbound or outbound).
```